src/main/drivers/flash_w25n.c

```diff
@@ -0,0 +1,390 @@
+{
+    w25n_enable(bus);
+    spiTransfer(bus->busdev_u.spi.instance, txData, rxData, len);
+    w25n_disable(bus);
+}
+
+/**
+ * Send the given command byte to the device.
+ */
+static void w25n_performOneByteCommand(busDevice_t *bus, uint8_t command)
+{
+    w25n_enable(bus);
+
+    spiTransferByte(bus->busdev_u.spi.instance, command);
+
+    w25n_disable(bus);
+}
+
+/**
+ * The flash requires this write enable command to be sent before commands that would cause
+ * a write like program and erase.
+ */
+static void w25n_writeEnable(flashDevice_t *fdevice)
+{
+    w25n_performOneByteCommand(fdevice->busdev, W25N_INSTRUCTION_WRITE_ENABLE);
+
+    // Assume that we're about to do some writing, so the device is just about to become busy
+    fdevice->couldBeBusy = true;
+}
+
+static uint8_t w25n_readStatus(busDevice_t *bus)
+{
+    const uint8_t command[3] = { W25N_INSTRUCTION_READ_STATUS_REG, W25N_STATUS_Register_3, 0 };
+    uint8_t in[3];
+
+    w25n_transfer(bus, command, in, sizeof(command));
+
+    return in[2];
+}
+
+static bool w25n_isReady(flashDevice_t *fdevice)
+{
+    // If couldBeBusy is false, don't bother to poll the flash chip for its status
+    fdevice->couldBeBusy = fdevice->couldBeBusy && ((w25n_readStatus(fdevice->busdev) & W25N_STATUS_FLAG_WRITE_ENABLE_LATCH) != 0);
+
+    return !fdevice->couldBeBusy;
+}
+
+static bool w25n_waitForReady(flashDevice_t *fdevice, uint32_t timeoutMillis)
+{
+    uint32_t time = millis();
+    while (!w25n_isReady(fdevice)) {
+        if (millis() - time > timeoutMillis) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ * Read chip identification and geometry information (into global `geometry`).
+ *
+ * Returns true if we get valid ident, false if something bad happened like there is no W25N.
+ */
+
+bool w25n_detect(flashDevice_t *fdevice, uint32_t chipID)
+{
+    switch (chipID) {
+    case JEDEC_ID_WINBOND_W25N01GV:
+        fdevice->geometry.sectors = 1024;
+        fdevice->geometry.pagesPerSector = 64;
+        break;
+    default:
+        // Unsupported chip or not an SPI NAND flash
+        fdevice->geometry.sectors = 0;
+        fdevice->geometry.pagesPerSector = 0;
+        fdevice->geometry.sectorSize = 0;
+        fdevice->geometry.totalSize = 0;
+        return false;
+    }
+
+    fdevice->geometry.flashType = FLASH_TYPE_NAND;
+    fdevice->geometry.pageSize = W25N_PAGESIZE;
+    fdevice->geometry.sectorSize = fdevice->geometry.pagesPerSector * fdevice->geometry.pageSize;
+    fdevice->geometry.totalSize = fdevice->geometry.sectorSize * fdevice->geometry.sectors;
+
+    fdevice->couldBeBusy = true; // Just for luck we'll assume the chip could be busy even though it isn't specced to be
+    fdevice->vTable = &w25n_vTable;
+    spiSetDivisor(fdevice->busdev->busdev_u.spi.instance, SPI_CLOCK_ULTRAFAST);
+    return true;
+}
+
+static void w25n_setPageAddress(uint8_t *buf, uint32_t address)
+{
+    *buf++ = (address >> 20) & 0xff;
```

Comment:
I would use bitfields instead. We already do that for SBUS decoding and like

Reply:
Need to investigate.